awscli/customizations/s3/fileinfo.py

```diff
@@ -48,22 +45,26 @@ def save_file(filename, response_data, last_update, is_stream=False):
             if not e.errno == errno.EEXIST:
                 raise CreateDirectoryError(
                     "Could not create directory %s: %s" % (d, e))
-    md5 = hashlib.md5()
+
+    if MD5_AVAILABLE and _can_validate_md5_with_etag(etag, response_data):
+        md5 = hashlib.md5()
+    else:
+        md5 = None
+
     file_chunks = iter(partial(body.read, 1024 * 1024), b'')
     if is_stream:
         # Need to save the data to be able to check the etag for a stream
         # because once the data is written to the stream there is no
         # undoing it.
-        payload = write_to_file(None, etag, md5, file_chunks, True)
+        payload = write_to_file(None, etag, file_chunks, md5, True)
     else:
         with open(filename, 'wb') as out_file:
-            write_to_file(out_file, etag, md5, file_chunks)
+            write_to_file(out_file, etag, file_chunks, md5)
 
-    if _can_validate_md5_with_etag(etag, response_data):
-        if etag != md5.hexdigest():
-            if not is_stream:
-                os.remove(filename)
-            raise MD5Error(filename)
+    if md5 and etag != md5.hexdigest():
```

Comment:
Yes.  Semantically, if you have a value of `None`, you need to differentiate between "provided a false like value" and "did not provide a value".